### PR TITLE
[Workaround Only, Will Revert] Add Topology parameter to NodeName lines for block topology nodes

### PR DESCRIPTION
**Pull request by hehe7318**
### Description of changes
Workaround for Slurm 25.11.2 bug where topology_p_add_rm_node incorrectly removes nodes from topology blocks when they enter power-down state. By setting `Topology=default:<BlockName>` on NodeName lines, the node's `topology_str` is no longer NULL, preventing the buggy code path.

SchedMD has confirmed this is a known issue with a fix in review.

## Changes we are doing

The change in this PR is to read the topology.conf that ParallelCluster creates and stores at `/opt/slurm/etc/` and parse it to get Block Mappings.
So if this file exists we read it to find the BlockName so that we can add `Topology=default:<BlockName>` to the slurm node definition

## Why not create the topology.conf through pcluster_slurm_config_generator.py or merge the configuration in file?
This is a temporary workaround for the bug that we found with slurm and we do not change the generation of separate topology.conf file since in future when we want to expand to use different topology plugin we can easily expand the existing resource we have in our cookbook.

### Tests
* Running

### References
* Link to impacted open issues.
* Link to related PRs in other packages (i.e. cookbook, node).
* Link to documentation useful to understand the changes.

### Checklist
- Make sure you are pointing to **the right branch**.
- If you're creating a patch for a branch other than `develop` add the branch name as prefix in the PR title (e\.g\. `[release-3.6]`).
- Check all commits' messages are clear, describing what and why vs how.
- Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by himani2411** [comment]
Added unit tests in https://github.com/aws/aws-parallelcluster-cookbook/pull/3123

File: cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py [hunk 1]
```diff
@@ -40,6 +40,60 @@ class CriticalError(Exception):
...
+    if not path.exists(topology_conf_path):
+        return mapping
+
+    try:
+        with open(topology_conf_path, encoding="utf-8") as f:
+            for line in f:
+                line = line.strip()
+                if not line or line.startswith("#"):
+                    continue
+                if not line.startswith("BlockName="):
```
Comment (gmarciani) [inline cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:61]: What about lines starting with `BlockSizes` or `Nodes`?
REf https://slurm.schedmd.com/topology.conf.html

Instead of skipping lines that do not start with `BlockName=`, we should instead consider all the lines containing both `BlockName=` and `Nodes=` because ultimately those are the info we want to parse to build the mapping

File: cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py [hunk 2]
```diff
@@ -40,6 +40,60 @@ class CriticalError(Exception):
...
+                    continue
+
+                block_name = None
+                nodes_spec = None
+                for token in line.split():
```
Comment (gmarciani) [inline cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:66]: [non blocking] Why looping+splitting rather than a regex to detect the values?

File: cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py [hunk 3]
```diff
@@ -40,6 +40,60 @@ class CriticalError(Exception):
...
+                        nodes_spec = token.split("=", 1)[1]
+
+                if not block_name or not nodes_spec:
+                    continue
+
+                # Extract queue_name and compute_resource_name from node spec
+                # Node spec format: {queue_name}-st-{compute_resource_name}-[1-N]
+                #                or {queue_name}-dy-{compute_resource_name}-[1-N]
+                # Remove the trailing -[1-N] part
+                node_prefix = re.sub(r"-\[.*\]$", "", nodes_spec)
```
Comment (gmarciani) [inline cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:79]: Is this logic able to handle the case of queue name and cr name containing `-st|dy-`?

The extraction of queue name and cr name from a nodename deserves a dedicated function with a dedicated unit test to validate the robustness of the regex in parsing edge cases.

File: cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py (hunk 1 above)
Comment (himani2411) [reply cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:61]: As per the slurm documentation you referred, the lines should start with BlockName or BlockSizes and they should be on separate lines.

1. IF the line start with Nodes it will fail slurm validation with an error `error: _parse_next_key: Parsing error at unrecognized key: Nodes`
2. And if the line start with BlockSizes with other details in the same line it gives error `error: Parse error in file /opt/slurm/etc/topology.conf line 3: "  BlockName=Block1  Nodes=q1-st-cr1-[1-2]"`

File: cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py (hunk 3 above)
Comment (himani2411) [reply cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:79]: Will add the unit test for verifying this use-case, anyway might have to do these changes for code-linters too as the function is too complex

Comment (himani2411) [reply cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:79]: However the use of `dy` is an overkill since Block topology we use is only with static nodes/Capacity blocks and only for p6e-gb200 instance.

Comment (himani2411) [reply cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm/pcluster_slurm_config_generator.py:79]: we will handle the workaround + unit tests in https://github.com/aws/aws-parallelcluster-cookbook/pull/3123 instead of using the fragile regex-based topology.conf parsing in pcluster_slurm_config_generator.py which was the old approach in this PR.

**Comment by himani2411** [comment]
Closing as fix from slurm is in v25.11.3 https://github.com/SchedMD/slurm/releases/tag/slurm-25-11-3-1
